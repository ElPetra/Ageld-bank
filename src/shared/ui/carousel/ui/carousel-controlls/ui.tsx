import cn from 'classnames';
import { type RefObject } from 'react';
import { useCarouselControlls } from 'src/shared/lib/carousel';
import './styles.scss';
import { useTranslation } from 'react-i18next';

interface Props {
    containerRef: RefObject<HTMLDivElement>;
    angle: number;
}

export const Controlls = ({ containerRef, angle }: Props) => {
    const { changeRotateByArrow, changeRotateByControl, active } =
        useCarouselControlls(containerRef, angle);
    const { t } = useTranslation();
    const controllClasses = (el: 'center' | 'prev' | 'next') =>
        cn({
            controll: true,
            active_controll: el === active
        });
    return (
        <div className='carousel_controlls_container'>
            <div className='controll_block_container'>
                <div
                    className='controll_block'
                    onClick={() => changeRotateByControl('prev')}
                >
                    <div
                        className={controllClasses('prev')}
                        aria-label={t('предыдущий слайд')}
                        role='button'
                    ></div>
                </div>
                <div
                    className='controll_block'
                    onClick={() => changeRotateByControl('center')}
                >
                    <div
                        className={controllClasses('center')}
                        aria-label={t('актуальный слайд')}
                        role='button'
                    ></div>
                </div>
                <div
                    className='controll_block'
                    onClick={() => changeRotateByControl('next')}
                >
                    <div
                        className={controllClasses('next')}
                        aria-label={t('следующий слайд')}
                        role='button'
                    ></div>
                </div>
            </div>
            <div className='controll_block_container arrows'>
                <div
                    className='controll_block'
                    onClick={() => changeRotateByArrow('left')}
                >
                    <div
                        className='arrow controll'
                        aria-label={t('переключить назад')}
                        role='button'
                    ></div>
                </div>
                <div
                    className='controll_block right_arrow'
                    onClick={() => changeRotateByArrow('right')}
                >
                    <div
                        className='arrow controll'
                        aria-label={t('переключить вперед')}
                        role='button'
                    ></div>
                </div>
            </div>
        </div>
    );
};
