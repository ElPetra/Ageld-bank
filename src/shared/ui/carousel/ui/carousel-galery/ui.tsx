import cn from 'classnames';

import type { RefObject } from 'react';
import './styles.scss';

interface Props {
    containerRef: RefObject<HTMLDivElement>;
    distance: number;
    angle: number;
    cards: Array<JSX.Element>;
}

export const Gallery = ({ containerRef, distance, angle, cards }: Props) => {
    const innerContainerClasses = cn({
        carousel_inner_container: cards.length === 3,
        carousel_inner_container_hidden: cards.length > 3
    });
    return (
        <div className={innerContainerClasses}>
            <div className='carousel_container' ref={containerRef}>
                <div
                    className='carousel_element'
                    style={{ transform: `translateZ(${distance}px)` }}
                >
                    {cards[0]}
                </div>
                {distance &&
                    cards.map((el, i) => {
                        if (i !== 0) {
                            return (
                                <div
                                    key={i}
                                    className='carousel_element'
                                    style={{
                                        transform: `rotateY(${i * angle}deg) translateZ(${distance}px)`
                                    }}
                                >
                                    {el}
                                </div>
                            );
                        }
                        return null;
                    })}
            </div>
        </div>
    );
};
