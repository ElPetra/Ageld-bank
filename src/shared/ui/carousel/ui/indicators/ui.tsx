import cn from 'classnames';

import type { ReactNode } from 'react';

import './styles.scss';

interface Props {
    changeRotateByIndicator: (indicator: number) => void;
    active: number;
    cards: ReactNode[];
}

export const CarouselIndicators = ({
    changeRotateByIndicator,
    active,
    cards
}: Props) => {
    const indicatorClasses = (index: number) =>
        cn('carousel-indicators__indicator', {
            active: active == index
        });
    return (
        <div className='carousel-indicators'>
            {cards.map((_, index) => (
                <button
                    key={index}
                    className={indicatorClasses(index)}
                    onClick={() => changeRotateByIndicator(index)}
                />
            ))}
        </div>
    );
};
