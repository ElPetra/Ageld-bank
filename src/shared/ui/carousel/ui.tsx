import { useRef } from 'react';
import { useCarousel, useCarouselControls } from 'src/shared/lib/carousel';

import { CarouselInner } from './ui/inner';
import { CarouselIndicators } from './ui/indicators';
import { CarouselControls } from './ui/controls';

import type { ReactNode } from 'react';

import './styles.scss';

interface Props {
    cards: ReactNode[];
}

export const Carousel = ({ cards }: Props) => {
    const containerRef = useRef<HTMLDivElement>(null);
    const { angle, distance } = useCarousel(cards.length);
    const { changeRotateByIndicator, changeRotateByControl, active } =
        useCarouselControls(cards.length, containerRef, angle);
    return (
        <div className='carousel'>
            <div>
                <CarouselInner
                    distance={distance}
                    angle={angle}
                    active={active}
                    containerRef={containerRef}
                    cards={cards}
                />
                <CarouselIndicators
                    cards={cards}
                    changeRotateByIndicator={changeRotateByIndicator}
                    active={active}
                />
            </div>
            <CarouselControls changeRotateByControl={changeRotateByControl} />
        </div>
    );
};
