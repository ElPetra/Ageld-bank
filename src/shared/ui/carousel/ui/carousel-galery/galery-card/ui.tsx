import { Button, Card, Text } from 'src/shared/ui';
import './styles.scss';
import { type ReactNode } from 'react';

interface Props {
    title: string;
    content: string | ReactNode;
    button: string;
}

export const GalleryCard = ({ title = 'Новинка!', content, button }: Props) => {
    return (
        <Card
            className='gallery_card'
            justify='space-between'
            direction='column'
            gap='small'
            borderRadius='large'
            padding='medium'
        >
            <Text size='m' weight='bold' align='center'>
                {title}
            </Text>
            <Text size='m'>{content}</Text>
            <Button variant='secondary' size='medium' width='max'>
                {button}
            </Button>
        </Card>
    );
};
