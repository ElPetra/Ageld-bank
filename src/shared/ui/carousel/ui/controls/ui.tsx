import { Icon } from 'src/shared/ui';

import './styles.scss';

interface Props {
    changeRotateByControl: (arrow: 'prev' | 'next') => void;
}

export const CarouselControls = ({ changeRotateByControl }: Props) => {
    return (
        <div className='carousel-controls'>
            <button
                className='carousel-controls__prev'
                onClick={() => changeRotateByControl('prev')}
            >
                <Icon width={40} icon='arrow-button-left' />
            </button>
            <button
                className='carousel-controls__next'
                onClick={() => changeRotateByControl('next')}
            >
                <Icon width={40} icon='arrow-button-right' />
            </button>
        </div>
    );
};
