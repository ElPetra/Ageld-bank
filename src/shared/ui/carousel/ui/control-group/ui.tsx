import { useCarouselControls } from 'src/shared/lib';

import { CarouselIndicators } from './indicators';
import { CarouselControls } from './controls';

import type { ReactNode, RefObject } from 'react';

interface Props {
    cards: ReactNode[];
    angle: number;
    containerRef: RefObject<HTMLDivElement>;
}

export const CarouselControlGroup = ({ cards, containerRef, angle }: Props) => {
    const { changeRotateByIndicator, changeRotateByControl, active } =
        useCarouselControls(cards.length, containerRef, angle);
    return (
        <>
            <CarouselIndicators
                cards={cards}
                changeRotateByIndicator={changeRotateByIndicator}
                active={active}
            />
            <CarouselControls changeRotateByControl={changeRotateByControl} />
        </>
    );
};
