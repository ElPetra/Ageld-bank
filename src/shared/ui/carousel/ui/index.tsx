import { useRef } from 'react';

import { useCarousel } from '../lib';

import { CarouselControlGroup } from './control-group';
import { CarouselInner } from './inner';

import type { ReactNode } from 'react';

import '../styles.scss';

interface Props {
    cards: ReactNode[];
}

export const Carousel = ({ cards }: Props) => {
    const containerRef = useRef<HTMLDivElement>(null);
    const { angle, distance } = useCarousel(cards.length);
    return (
        <div className='carousel' data-testid='carousel'>
            <div>
                <CarouselInner
                    distance={distance}
                    angle={angle}
                    containerRef={containerRef}
                    cards={cards}
                />
            </div>
            <CarouselControlGroup
                cards={cards}
                angle={angle}
                containerRef={containerRef}
            />
        </div>
    );
};
