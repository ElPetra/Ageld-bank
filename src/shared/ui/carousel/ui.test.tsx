import { render, screen, renderHook } from '@testing-library/react';

import '@testing-library/jest-dom';
import '@testing-library/jest-dom/jest-globals';

import { Carousel } from './ui';
import { useCarousel } from './lib';

describe('Carousel ui', () => {
    test('match snapshot', () => {
        render(
            <Carousel
                cards={[
                    <div key={1}>Карточка 1</div>,
                    <div key={2}>Карточка 2</div>,
                    <div key={3}>Карточка 3</div>,
                    <div key={4}>Карточка 4</div>,
                    <div key={5}>Карточка 5</div>
                ]}
            />
        );
        const carousel = screen.getByTestId('carousel');
        expect(carousel).toMatchSnapshot();
    });

    test('Carousel renders', () => {
        render(
            <Carousel
                cards={[
                    <div key={1}>Карточка 1</div>,
                    <div key={2}>Карточка 2</div>,
                    <div key={3}>Карточка 3</div>,
                    <div key={4}>Карточка 4</div>,
                    <div key={5}>Карточка 5</div>
                ]}
            />
        );
        const carousel = screen.getByTestId('carousel');
        expect(carousel).toBeInTheDocument();
    });
});

describe('calculate the rotation angle and distance from the center', () => {
    it('should calculate correct angle and distance for correct length', () => {
        const length = 4;
        const { result } = renderHook(() => useCarousel(length));
        const { angle, distance } = result.current;
        const edge = 500 / Math.sin((180 / length) * (Math.PI / 180)) / 2;
        const height = Math.sqrt(4 * edge ** 2 - 500 ** 2) / 2;
        expect(angle).toBe(90);
        expect(distance).toBe(height);
    });
});
