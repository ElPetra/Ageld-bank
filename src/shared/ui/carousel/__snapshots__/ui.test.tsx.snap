// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel ui match snapshot 1`] = `
<div
  class="carousel"
  data-testid="carousel"
>
  <div>
    <div
      class="carousel-inner overflowHidden"
    >
      <div
        class="carousel-inner__container"
      >
        <div
          class="carousel-inner__element"
          style="transform: rotateY(0deg) translateZ(0px);"
        >
          <div>
            Карточка 1
          </div>
        </div>
        <div
          class="carousel-inner__element"
          style="transform: rotateY(72deg) translateZ(0px);"
        >
          <div>
            Карточка 2
          </div>
        </div>
        <div
          class="carousel-inner__element"
          style="transform: rotateY(144deg) translateZ(0px);"
        >
          <div>
            Карточка 3
          </div>
        </div>
        <div
          class="carousel-inner__element"
          style="transform: rotateY(216deg) translateZ(0px);"
        >
          <div>
            Карточка 4
          </div>
        </div>
        <div
          class="carousel-inner__element"
          style="transform: rotateY(288deg) translateZ(0px);"
        >
          <div>
            Карточка 5
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="carousel-indicators"
  >
    <button
      class="carousel-indicators__indicator active"
    />
    <button
      class="carousel-indicators__indicator"
    />
    <button
      class="carousel-indicators__indicator"
    />
    <button
      class="carousel-indicators__indicator"
    />
    <button
      class="carousel-indicators__indicator"
    />
  </div>
  <div
    class="carousel-controls"
  >
    <button
      class="carousel-controls__prev"
    >
      <img
        alt=""
        height="24"
        width="40"
      />
    </button>
    <button
      class="carousel-controls__next"
    >
      <img
        alt=""
        height="24"
        width="40"
      />
    </button>
  </div>
</div>
`;
